File: src/component/hud/TownHUD.java
package component.hud;

import java.util.Queue;

import org.newdawn.slick.Color;
import org.newdawn.slick.Font;
import org.newdawn.slick.gui.ComponentListener;
import org.newdawn.slick.gui.GUIContext;

import component.Button;
import component.Component;
import component.Label;
import component.Panel;
import component.Label.VerticalAlignment;
import component.sprite.Sprite;

import core.ConstantStore;
import core.FontStore;
import core.ImageStore;

/**
 * A HUD holds quick information for the player in {@code TownScene}.
 */
public class TownHUD extends HUD {
	private static final int MARGIN = 10;
	
	private static final int BUTTON_HEIGHT = HEIGHT - (2 * MARGIN);
	
	private static final int INFO_WIDTH = 200;
	
	private Button trailButton;
	
	private Label timeLabel;
	private Label dateLabel;
	private Label notificationLabel;
	
	private Queue<String> notificationQueue;
	
	/**
	 * Constructs a HUD with a {@code GUIContext} and {@code HUDDataSource}.
	 * @param context Context
	 */
	public TownHUD(GUIContext context, ComponentListener listener) {
		super(context, context.getWidth(), HEIGHT);
		
		int panelWidth = context.getWidth() - INFO_WIDTH - MARGIN * 2;
		
		Panel panel = new Panel(container, panelWidth, HEIGHT);
	
		Font fieldFont = FontStore.get().getFont(FontStore.FontID.FIELD);
		
		Label menuLabel = new Label(container, fieldFont, Color.white, ConstantStore.get("TOWN_SCENE", "TRAIL"));
		
		Sprite fireSprite = new Sprite(container, 48, ImageStore.get().getImage("CAMP_ICON"));
		
		trailButton = new Button(container, menuLabel.getWidth() + (2 * MARGIN), BUTTON_HEIGHT, menuLabel);
		trailButton.setSprite(fireSprite);
		trailButton.setShowLabel(false);
		trailButton.addListener(listener);
		panel.add(trailButton, getPosition(ReferencePoint.TOPLEFT), ReferencePoint.TOPLEFT, MARGIN, MARGIN);
		
		int notificationWidth = panelWidth - trailButton.getWidth() -  MARGIN * 2;
		
		notificationLabel = new Label(container, notificationWidth, BUTTON_HEIGHT, fieldFont, Color.white, "");
		notificationLabel.setVerticalAlignment(VerticalAlignment.CENTER);
		notificationLabel.setBackgroundColor(Color.black);
		panel.add(notificationLabel, trailButton.getPosition(ReferencePoint.TOPRIGHT), ReferencePoint.TOPLEFT, MARGIN, 0);
		
		add(panel, getPosition(ReferencePoint.TOPLEFT), ReferencePoint.TOPLEFT);
		
		timeLabel = new Label(context, INFO_WIDTH, fieldFont.getLineHeight(), fieldFont, Color.white, "");
		add(timeLabel, notificationLabel.getPosition(ReferencePoint.CENTERRIGHT), ReferencePoint.BOTTOMLEFT, MARGIN, - MARGIN / 2);
		
		dateLabel = new Label(context, INFO_WIDTH, fieldFont.getLineHeight(), fieldFont, Color.white, "");
		add(dateLabel, notificationLabel.getPosition(ReferencePoint.CENTERRIGHT), ReferencePoint.TOPLEFT, MARGIN, MARGIN / 2);
		
		setBackgroundColor(Color.gray);
		setBevelWidth(2);
		setBevel(Component.BevelType.OUT);
		setBottomBorderWidth(2);
		setBorderColor(Color.black);
	}
	
	/**
	 * Update the party information on the screen (date and money).
	 */
	public void updatePartyInformation(String time, String date) {
		setTime(time);
		setDate(date);
	}
	
	/**
	 * Set the date label's contents.
	 * @param time New string from date label
	 */
	public void setTime(String time) {
		timeLabel.setText(time);
	}
	
	/**
	 * Set the date label's contents.
	 * @param date Amount of money to display
	 */
	public void setDate(String date) {
		dateLabel.setText(date);
	}
	
	public boolean isNotificationsEmpty() {
		return notificationQueue.isEmpty();
	}
	
	/**
	 * Set the notification for the notification bar.
	 * @param message New message to show
	 */
	public void setNotification(String message) {
		notificationLabel.setText(message);
	}
	
	/**
	 * Get the trail button.
	 * @return Trail button
	 */
	public Button getTrailButton() {
		return trailButton;
	}
}

File: web/src/component/hud/TownHUD.ts
import { HUD } from './HUD';
import { Panel } from '../Panel';
import { Button, ButtonClickListener } from '../Button';
import { Label, VerticalAlignment } from '../Label';
import { Sprite } from '../Sprite';
import { BevelType, ReferencePoint } from '../Component';
import { Color } from '../../core/Color';
import { FontStore, FontID } from '../../core/FontStore';
import { ImageStore } from '../../core/ImageStore';
import { get as getLiteral } from '../../core/ConstantStore';

export class TownHUD extends HUD {
  private static readonly MARGIN = 10;
  private static readonly BUTTON_HEIGHT = HUD.HEIGHT - 2 * TownHUD.MARGIN;
  private static readonly INFO_WIDTH = 200;

  private trailButton!: Button;
  private timeLabel!: Label;
  private dateLabel!: Label;
  private notificationLabel!: Label;
  private notificationQueue: string[] = [];

  constructor(width: number, listener: ButtonClickListener) {
    super(width, HUD.HEIGHT);

    const panelWidth = width - TownHUD.INFO_WIDTH - TownHUD.MARGIN * 2;
    const panel = new Panel(panelWidth, HUD.HEIGHT);

    const fieldFont = FontStore.getFont(FontID.FIELD);

    const menuLabel = Label.withTextWidth(fieldFont, Color.white, getLiteral('TOWN_SCENE', 'TRAIL'));

    const trailImage = ImageStore.getImage('TRAIL_ICON');
    const trailSprite = new Sprite(48);
    trailSprite.setImage(trailImage);

    this.trailButton = new Button(
      menuLabel.getWidth() + 2 * TownHUD.MARGIN,
      TownHUD.BUTTON_HEIGHT,
      menuLabel
    );
    this.trailButton.setSprite(trailSprite);
    this.trailButton.setShowLabel(false);
    this.trailButton.addClickListener(listener);
    panel.add(
      this.trailButton,
      this.getPosition(ReferencePoint.TOPLEFT),
      ReferencePoint.TOPLEFT,
      TownHUD.MARGIN,
      TownHUD.MARGIN
    );

    this.trailButton.setTooltipEnabled(true);
    this.trailButton.setTooltipMessage(getLiteral('TOWN_SCENE', 'CONTINUE_ON_TRAIL'));

    const notificationWidth = panelWidth - this.trailButton.getWidth() - TownHUD.MARGIN * 2;

    this.notificationLabel = new Label(
      notificationWidth,
      TownHUD.BUTTON_HEIGHT,
      fieldFont,
      Color.white,
      ''
    );
    this.notificationLabel.setVerticalAlignment(VerticalAlignment.CENTER);
    this.notificationLabel.setBackgroundColor(Color.black);
    panel.add(
      this.notificationLabel,
      this.trailButton.getPosition(ReferencePoint.TOPRIGHT),
      ReferencePoint.TOPLEFT,
      TownHUD.MARGIN,
      0
    );

    this.add(panel, this.getPosition(ReferencePoint.TOPLEFT), ReferencePoint.TOPLEFT);

    this.timeLabel = new Label(
      TownHUD.INFO_WIDTH,
      fieldFont.getLineHeight(),
      fieldFont,
      Color.white,
      ''
    );
    this.add(
      this.timeLabel,
      this.notificationLabel.getPosition(ReferencePoint.CENTERRIGHT),
      ReferencePoint.BOTTOMLEFT,
      TownHUD.MARGIN,
      -TownHUD.MARGIN / 2
    );

    this.dateLabel = new Label(
      TownHUD.INFO_WIDTH,
      fieldFont.getLineHeight(),
      fieldFont,
      Color.white,
      ''
    );
    this.add(
      this.dateLabel,
      this.notificationLabel.getPosition(ReferencePoint.CENTERRIGHT),
      ReferencePoint.TOPLEFT,
      TownHUD.MARGIN,
      TownHUD.MARGIN / 2
    );

    this.setBackgroundColor(Color.gray);
    this.setBevelWidth(2);
    this.setBevel(BevelType.OUT);
    this.setBottomBorderWidth(2);
    this.setBorderColor(Color.black);
  }

  updatePartyInformation(time: string, date: string): void {
    this.setTime(time);
    this.setDate(date);
  }

  setTime(time: string): void {
    this.timeLabel.setText(time);
  }

  setDate(date: string): void {
    this.dateLabel.setText(date);
  }

  isNotificationsEmpty(): boolean {
    return this.notificationQueue.length === 0;
  }

  setNotification(message: string): void {
    this.notificationLabel.setText(message);
  }

  getTrailButton(): Button {
    return this.trailButton;
  }
}
